### PR TITLE
Created CustomUser model and added email and password validation

### PR DESCRIPTION
**Pull request by olexiyko**

### CODE REVIEW

File: forum/users/models.py [hunk 1]
```diff
@@ -1,3 +1,65 @@
...
+            raise ValueError("The Email field must be set")
+        if password:
+            password_pattern = r"^(?=.*[0-9])(?=.*[a-z])(?=.*[A-Z]).{8,}$"
+            if not re.match(password_pattern, password):
+                raise ValueError("Password is not matching valid pattern")
+        email = self.normalize_email(email)
+        user = self.model(email=email, **extra_fields)
+        user.set_password(password)
+        user.save(using=self._db)
+        return user
```
Comment (mehalyna) [inline forum/users/models.py:35]: Validating `email` and `password` patterns directly in the `CustomUserManager` tightly couples validation logic with the model. This is better handled in serializers or forms.
Move validation logic to a dedicated serializer. This keeps your model focused on database representation and behavior.

File: forum/users/models.py [hunk 2]
```diff
@@ -1,3 +1,65 @@
...
+    (0, "Unassigned"),
+    (1, "Startup"),
+    (2, "Investor"),
+]
+
+
+class CustomUserManager(BaseUserManager):
+    def create_user(self, email, password=None, **extra_fields):
+        if email:
+            email_pattern = r"^[a-zA-Z0-9]+@[a-zA-Z0-9]+\.[a-zA-Z0-9]+$"
```
Comment (mehalyna) [inline forum/users/models.py:22]: The regex for email validation is simplistic and might not cover edge cases.
You may use Django's built-in email validation:
```
from django.core.validators import validate_email
from django.core.exceptions import ValidationError

try:
    validate_email(email)
except ValidationError:
    raise ValueError("The Email is not valid")
```

File: forum/users/models.py [hunk 3]
```diff
@@ -1,3 +1,65 @@
...
+        extra_fields.setdefault("is_superuser", True)
+        return self.create_user(email, password, **extra_fields)
+
+
+class CustomUser(AbstractBaseUser, PermissionsMixin):
+    user_id = models.AutoField(primary_key=True)
+    first_name = models.CharField(max_length=30)
+    last_name = models.CharField(max_length=30)
+    email = models.EmailField(unique=True)
+    password = models.CharField(max_length=128)
```
Comment (mehalyna) [inline forum/users/models.py:48]: The `password` field is explicitly defined, which may cause redundancy with `AbstractBaseUser`.
It'll be better to remove the `password` field from the model since `AbstractBaseUser` already provides it. Instead, use `set_password` to handle hashing securely.

File: forum/users/models.py [hunk 4]
```diff
@@ -1,3 +1,65 @@
...
+    user_phone = models.CharField(max_length=30)
```
Comment (mehalyna) [inline forum/users/models.py:49]: The `user_phone` field is a simple `CharField`, which may lead to inconsistencies in phone number storage.
You may use the `PhoneNumberField` from the `django-phonenumber-field` package for better validation and consistency:
```
from phonenumber_field.modelfields import PhoneNumberField

user_phone = PhoneNumberField(null=True, blank=True)
```

File: forum/users/models.py [hunk 5]
```diff
@@ -1,3 +1,65 @@
...
+        extra_fields.setdefault("is_staff", True)
+        extra_fields.setdefault("is_superuser", True)
+        return self.create_user(email, password, **extra_fields)
+
+
+class CustomUser(AbstractBaseUser, PermissionsMixin):
+    user_id = models.AutoField(primary_key=True)
+    first_name = models.CharField(max_length=30)
+    last_name = models.CharField(max_length=30)
+    email = models.EmailField(unique=True)
```
Comment (mehalyna) [inline forum/users/models.py:47]: Add `db_index=True` to the `email` field for performance optimization

File: forum/users/models.py [hunk 6]
```diff
@@ -1,3 +1,65 @@
...
+
+    USERNAME_FIELD = "email"
+    REQUIRED_FIELDS = ["first_name", "last_name"]
+
+    class Meta:
+        db_table = "users"
+        ordering = ["user_id"]
+
+    def __str__(self):
+        return f"id:{self.user_id} - {self.first_name} - {self.last_name}"
```
Comment (mehalyna) [inline forum/users/models.py:65]: Include the role in the string representation:
```
def __str__(self):
    return f"id:{self.user_id} - {self.first_name} {self.last_name} (Role: {self.get_role_display()})"
```

File: forum/users/models.py [hunk 7]
```diff
@@ -1,3 +1,65 @@
...
+    last_login = models.DateTimeField(auto_now=True)
+
+    objects = CustomUserManager()
+
+    USERNAME_FIELD = "email"
+    REQUIRED_FIELDS = ["first_name", "last_name"]
+
+    class Meta:
+        db_table = "users"
+        ordering = ["user_id"]
```
Comment (mehalyna) [inline forum/users/models.py:62]: Add a `verbose name` and `name_plural` for better admin readability:
```
class Meta:
    db_table = "users"
    ordering = ["user_id"]
    verbose_name = "User"
    verbose_name_plural = "Users"
```

File: forum/users/models.py
Comment (mehalyna) [inline forum/users/models.py:1]: Fields like `user_phone` and `title` are required by default, which might not be intended.
Add `null=True`, `blank=True` where appropriate.
Example:
```
user_phone = models.CharField(max_length=30, null=True, blank=True)
title = models.CharField(max_length=30, null=True, blank=True)
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/validators.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
+from django.core.exceptions import ValidationError
+from rest_framework import serializers
+
+
+class SpecialCharacterValidator:
+    def validate(self, password, user=None):
+        if not any(char in "!@#$%^&*()_+" for char in password):
```
Comment (mehalyna) [inline forum/users/validators.py:7]: The special characters list is hardcoded, which limits flexibility. Allow the character set to be customizable via a class attribute:
```
class SpecialCharacterValidator:
    special_characters = "!@#$%^&*()_+"

    def validate(self, password, user=None):
        if not any(char in self.special_characters for char in password):
            raise ValidationError(
                f"The password must contain at least one special character ({self.special_characters})."
            )
```

File: forum/users/validators.py
Comment (mehalyna) [inline forum/users/validators.py:1]: Validators should be tested to ensure they function as expected.
Write unit tests using Django’s `TestCase`

File: forum/users/validators.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        if not any(char.isupper() for char in password):
+            raise serializers.ValidationError(
+                {"password": "The password must contain at least one upper character"}
+            )
+
+    def get_help_text(self):
+        return (
+            "Your password must contain at least one special character (!@#$%^&*()_+) "
+            "and one uppercase letter."
+        )
```
Comment (mehalyna) [inline forum/users/validators.py:22]: Format the message for better readability.

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/validators.py (hunk 9 above)
Comment (olexiyko) [reply forum/users/validators.py:22]: In this case i used a plagin BlackFormatter for better readability.Should I change it to another?

File: forum/users/models.py [hunk 10]
```diff
@@ -1,3 +1,57 @@
...
+import datetime
+import re
+from phonenumber_field.modelfields import PhoneNumberField
 
-# Create your models here.
+ROLE_CHOICES = [
+    (0, "Unassigned"),
+    (1, "Startup"),
+    (2, "Investor"),
+]
```
Comment (devdsys) [inline forum/users/models.py:15]: At the team meeting on December 16, it was decided to add the field `Both` to `ROLE_CHOICES`:  

```python
ROLE_CHOICES = [
    (0, "Unassigned"),
    (1, "Startup"),
    (2, "Investor"),
    (3, "Both")
]
```

**Review by devdsys — CHANGES_REQUESTED** [review]
Please make the requested changes to the code based on the comments.

File: forum/users/models.py [hunk 11]
```diff
@@ -1,3 +1,59 @@
...
+)
+from phonenumber_field.modelfields import PhoneNumberField
 
-# Create your models here.
+ROLE_CHOICES = [
+    (0, "Unassigned"),
+    (1, "Startup"),
+    (2, "Investor"),
+    (3, "Both")
+]
```
Comment (mehalyna) [inline forum/users/models.py:14]: Just the recommendation, - replace the tuple with an `Enum` for better readability and maintainability.
```
from enum import IntEnum

class Role(IntEnum):
    UNASSIGNED = 0
    STARTUP = 1
    INVESTOR = 2
    BOTH = 3

ROLE_CHOICES = [(role.value, role.name.capitalize()) for role in Role]
```

This makes it easier to reference roles in code:

```
if user.role == Role.STARTUP:
```

File: forum/users/serializers.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+            setattr(instance, attr, value)
+        if password:
+            instance.set_password(password)
+        instance.save()
+        return instance
+
+    def validate_email(self, value):
+        if CustomUser.objects.filter(email=value).exists():
+            raise serializers.ValidationError("A user with this email already exists.")
+        return value
```
Comment (mehalyna) [inline forum/users/serializers.py:45]: You are checking for existing users with the same email during validation, but this could cause conflicts when updating a user. Add logic to exclude the current user during updates.

File: forum/users/serializers.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        return instance
+
+    def update(self, instance, validated_data):
+        password = validated_data.pop("password", None)
+        for attr, value in validated_data.items():
+            setattr(instance, attr, value)
+        if password:
+            instance.set_password(password)
+        instance.save()
+        return instance
```
Comment (mehalyna) [inline forum/users/serializers.py:40]: When updating an instance, you should ensure that `set_password` is only used for password updates. The current logic works but might accidentally expose passwords in logs if misconfigured.
You may add explicit password handling

File: forum/users/serializers.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+        ]
+        read_only_fields = ("created_at", "updated_at")
+        extra_kwargs = {"password": {"write_only": True}}
+
+    def create(self, validated_data):
+        password = validated_data.pop("password")
+        instance = CustomUser.objects.create(**validated_data)
+        instance.set_password(password)
+        instance.save()
+        return instance
```
Comment (mehalyna) [inline forum/users/serializers.py:31]: When no password is provided during creation, `set_password` might raise an issue. Ensure a proper check before saving.

File: forum/users/tests.py [hunk 15]
```diff
@@ -1,3 +1,45 @@
...
+                except ValidationError:
+                    self.fail(
+                        f"Valid password '{password}' raised ValidationError unexpectedly."
+                    )
+
+    def test_invalid_passwords(self):
+        for invalid_password in self.invalid_password_cases:
+            with self.subTest(invalid_password=invalid_password):
+                with self.assertRaises(ValidationError):
+                    self.validator.validate(invalid_password)
```
Comment (mehalyna) [inline forum/users/tests.py:45]: When testing invalid passwords, you can assert not only that a `ValidationError` is raised but also check the specific error message. This ensures the validator raises the correct feedback for each failure.
Example:
```
def test_invalid_passwords(self):
    for invalid_password in self.invalid_password_cases:
        with self.subTest(invalid_password=invalid_password):
            with self.assertRaises(ValidationError) as context:
                self.validator.validate(invalid_password)
            self.assertIn(
                "The password must contain at least one special character",
                str(context.exception),
                f"Invalid password '{invalid_password}' did not raise the correct error message."
            )
```

File: forum/users/tests.py
Comment (mehalyna) [inline forum/users/tests.py:1]: Add edge cases to test borderline conditions such as:
- Only one special character.
- Passwords with a mix of special characters.
-Very long or very short passwords.
Example edge cases:
```
def test_edge_case_passwords(self):
    edge_cases = [
        "!A1",  # Minimum valid password
        "A" * 50 + "@1",  # Long valid password
        "!!@#$%^&*()_+",  # Only special characters
    ]
    for password in edge_cases:
        with self.subTest(password=password):
            try:
                self.validator.validate(password)
            except ValidationError:
                self.fail(f"Edge case password '{password}' raised ValidationError unexpectedly.")
```

Comment (mehalyna) [inline forum/users/tests.py:1]: Explicitly check how the validator handles empty strings or `None` input. These cases could break the validator unexpectedly.
Example:
```
def test_empty_and_none_passwords(self):
    invalid_cases = ["", None]
    for password in invalid_cases:
        with self.subTest(password=password):
            with self.assertRaises(ValidationError):
                self.validator.validate(password)
```

**Review by mehalyna — CHANGES_REQUESTED** [review]

File: forum/users/serializers.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+            "email",
+            "user_phone",
+            "role",
+            "title",
+            "password",
+            "created_at",
+            "updated_at",
+        ]
+        read_only_fields = ("created_at", "updated_at")
+        extra_kwargs = {"password": {"write_only": True}}
```
Comment (IraMladanovych02) [inline forum/users/serializers.py:24]: please, set up min_length to extra_kwargs to validate the password (extra_kwargs = {"password": {"write_only": True, "min_length": 7}})

File: forum/users/serializers.py [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+            "last_name",
+            "email",
+            "user_phone",
+            "role",
+            "title",
+            "password",
+            "created_at",
+            "updated_at",
+        ]
+        read_only_fields = ("created_at", "updated_at")
```
Comment (IraMladanovych02) [inline forum/users/serializers.py:23]: add user_id to read_only_fields

**Review by IraMladanovych02 — APPROVED** [review]

**Review by mehalyna — APPROVED** [review]
